### PR TITLE
model: Add a parser for SPDX expressions

### PR DESCRIPTION
**Pull request by mnonnenmacher**
Implement a parser for SPDX expressions as specified in [1] using ANTLR
[2]. This is a prerequisite for implementing the evaluation of SPDX
expressions which will be used by upcoming features like evaluating the
compatibility of a license choice with an SPDX expression.

[1] https://spdx.org/spdx-specification-21-web-version#h.jxpfx0ykyb60
[2] https://www.antlr.org/

Signed-off-by: Martin Nonnenmacher <martin.nonnenmacher@here.com>

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1072)

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -3,6 +3,7 @@ dokkaPluginVersion = 0.9.17
 kotlinPluginVersion = 1.3.10
 versionsPluginVersion = 0.20.0
 
+antlrVersion = 4.7.1
```
Comment (fviernau) [inline gradle.properties]: Would it make sense to have a separate commit for integrating the library & plugin ?

File: model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+import io.kotlintest.shouldThrow
+import io.kotlintest.specs.WordSpec
+
+class SpdxExpressionParserTest : WordSpec() {
+    init {
+        "SpdxExpressionParser" should {
+            "parse a license id correctly" {
+                val spdxExpression = parseSpdxExpression("spdx.license-id")
+
+                spdxExpression shouldBe SpdxLicenseIdExpression("spdx.license-id", false)
```
Comment (fviernau) [inline model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt]: The 'false' value appears as 'magic number', would a named parameter still fit on the line?

File: model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+package com.here.ort.model.spdx
+
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldThrow
+import io.kotlintest.specs.WordSpec
+
+class SpdxExpressionParserTest : WordSpec() {
+    init {
+        "SpdxExpressionParser" should {
+            "parse a license id correctly" {
```
Comment (fviernau) [inline model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt]: For some of the test case names 'correctly' could be replaced with what correctly actually means, e.g. what is asserted...

File: model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+                        SpdxOperator.OR,
+                        SpdxCompoundExpression(
+                                SpdxLicenseIdExpression("license2", false),
+                                SpdxOperator.AND,
+                                SpdxLicenseIdExpression("license3", false)
+                        )
+                )
+            }
+
+            "bind the same operator in order" {
```
Comment (fviernau) [inline model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt]: how about: "and" should be "left associative" ?

File: model/src/main/antlr/com/here/ort/model/spdx/SpdxExpression.g4 [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    | OPEN compoundExpression CLOSE
+    ;
+
+licenseExpression
+    :
+    (simpleExpression | compoundExpression)
+    EOF
+    ;
+
+
```
Comment (sschuberth) [inline model/src/main/antlr/com/here/ort/model/spdx/SpdxExpression.g4]: Nit: Duplicate blank line.

File: model/src/main/antlr/com/here/ort/model/spdx/SpdxExpression.g4 [hunk 6]
```diff
@@ -0,0 +1,72 @@
+/*
```
Comment (sschuberth) [inline model/src/main/antlr/com/here/ort/model/spdx/SpdxExpression.g4]: Shouldn't we add our license / copyright header also to this source file?

**Comment by sschuberth** [comment]
All in all very nice... we should make a separate library out of it 😀

**Comment by mnonnenmacher** [comment]
@sschuberth I thought the same, there are a few more SPDX related features coming, once this is done it would make a lot of sense to move this to a standalone library which is released separately.

File: gradle.properties (hunk 1 above)
Comment (mnonnenmacher) [reply gradle.properties]: Probably, but I think in the past we usually introduced a dependency in the same commit as we started using it. It's only a few lines anyway, I think it makes more sense if the integration itself takes actual work beyond a very basic configuration.

File: model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt (hunk 2 above)
Comment (mnonnenmacher) [reply model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt]: Done. I have also made `false` the default value to get rid of the argument in most places.

File: model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt (hunk 3 above)
Comment (mnonnenmacher) [reply model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt]: If think the names can be significantly improved, mind making a PR later? I checked and couldn't really think of better names.

File: model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt (hunk 4 above)
Comment (mnonnenmacher) [reply model/src/test/kotlin/spdx/SpdxExpressionParserTest.kt]: Done.

File: model/src/main/antlr/com/here/ort/model/spdx/SpdxExpression.g4 (hunk 5 above)
Comment (mnonnenmacher) [reply model/src/main/antlr/com/here/ort/model/spdx/SpdxExpression.g4]: Done.

File: model/src/main/antlr/com/here/ort/model/spdx/SpdxExpression.g4 (hunk 6 above)
Comment (mnonnenmacher) [reply model/src/main/antlr/com/here/ort/model/spdx/SpdxExpression.g4]: Done.

File: gradle.properties (hunk 1 above)
Comment (fviernau) [reply gradle.properties]: yes makes sense, actually forgot the nit at the beginning - maybe not even that.

File: model/src/main/kotlin/spdx/SpdxExpressionDefaultVisitor.kt [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+class SpdxExpressionDefaultVisitor : SpdxExpressionBaseVisitor<SpdxExpression>() {
+    override fun visitLicenseExpression(ctx: SpdxExpressionParser.LicenseExpressionContext): SpdxExpression {
+        return when (ctx.childCount) {
+            2 -> visit(ctx.getChild(0))
+            else -> throw Error("SpdxExpression has invalid amount of children: '${ctx.childCount}'")
+        }
+    }
+
+    override fun visitCompoundExpression(ctx: SpdxExpressionParser.CompoundExpressionContext): SpdxExpression {
```
Comment (fviernau) [inline model/src/main/kotlin/spdx/SpdxExpressionDefaultVisitor.kt]: nit: you could import the 'CompoundExpressionContext' and also other innter classes like 'LicenseExpressionContext'
which gives a bit better readability. The names are unique enough to do that

File: model/src/main/kotlin/spdx/SpdxExpressionDefaultVisitor.kt [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            1 -> SpdxLicenseIdExpression(ctx.text)
+            2 -> SpdxLicenseIdExpression(ctx.text.dropLast(1), anyLaterVersion = true)
+            else -> throw Error("SpdxLicenseIdExpression has invalid amount of children: '${ctx.childCount}'")
+        }
+    }
+
+    override fun visitLicenseRefExpression(ctx: SpdxExpressionParser.LicenseRefExpressionContext): SpdxExpression {
+        return when (ctx.childCount) {
+            1 -> SpdxLicenseRefExpression(ctx.text)
+            else -> throw Error("SpdxLicenseRefExpression has invalid amount of children: '${ctx.childCount}'")
```
Comment (fviernau) [inline model/src/main/kotlin/spdx/SpdxExpressionDefaultVisitor.kt]: Is this quite generic type thrown out of your new component or is it always caught so that client code won't ever get it.
If it goes outside of the component I'd vote for introducing a new throwable type to allow for more fine grained handling.

File: model/src/test/kotlin/spdx/SpdxExpressionLexerTest.kt [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+import org.antlr.v4.runtime.BaseErrorListener
+import org.antlr.v4.runtime.CharStreams
+import org.antlr.v4.runtime.RecognitionException
+import org.antlr.v4.runtime.Recognizer
+
+class SpdxExpressionLexerTest : WordSpec() {
+    init {
+        "SpdxExpressionLexer" should {
+            "create the correct tokens for a valid expression" {
+                val testExpression = "(license-1 AND OR license.2 WITH 0LiCeNsE-3+) and (or with license-4 license-.5(("
```
Comment (fviernau) [inline model/src/test/kotlin/spdx/SpdxExpressionLexerTest.kt]: extracting e.g.: fun createLexerForExpression(expression: String): SpdxExpressionLexer
would remove some redundancy.

File: model/src/test/kotlin/spdx/SpdxExpressionLexerTest.kt [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                            recognizer: Recognizer<*, *>?,
+                            offendingSymbol: Any?,
+                            line: Int,
+                            charPositionInLine: Int,
+                            msg: String?,
+                            e: RecognitionException?
+                    ) {
+                        throw Error(msg)
+                    }
+                })
```
Comment (fviernau) [inline model/src/test/kotlin/spdx/SpdxExpressionLexerTest.kt]: Above I asked to extract a function to remove code redundancy. 
Seeing this I think you could extract lines 65 through 79 into a function:

fun getTokensForExpression(expression: String)

which would result in quite compact test cases.

File: model/src/main/kotlin/spdx/SpdxExpressionDefaultVisitor.kt (hunk 7 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/spdx/SpdxExpressionDefaultVisitor.kt]: Done.

File: model/src/main/kotlin/spdx/SpdxExpressionDefaultVisitor.kt (hunk 8 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/spdx/SpdxExpressionDefaultVisitor.kt]: Done. Added `SpdxException`.

File: model/src/test/kotlin/spdx/SpdxExpressionLexerTest.kt (hunk 9 above)
Comment (mnonnenmacher) [reply model/src/test/kotlin/spdx/SpdxExpressionLexerTest.kt]: Done as proposed below.

File: model/src/test/kotlin/spdx/SpdxExpressionLexerTest.kt (hunk 10 above)
Comment (mnonnenmacher) [reply model/src/test/kotlin/spdx/SpdxExpressionLexerTest.kt]: Done.
